packages/jsii/lib/assembler.ts

```diff
@@ -416,14 +393,18 @@ export class Assembler implements Emitter {
 
     if (this._isPrivateOrInternal(type.symbol)) {
       // Check if this type is "this" (explicit or inferred method return type).
-      const commonMessage = `cannot be used as the ${typeUse} because it is private or @internal`;
-      this._diagnostic(
-        typeAnnotationNode,
-        ts.DiagnosticCategory.Error,
-        isThisType
-          ? `Type "this" (aka: "${type.symbol.name}") ${commonMessage}`
-          : `Type "${type.symbol.name}" ${commonMessage}`,
-        makeCause(typeDeclaration),
+      this._diagnostics.push(
+        JsiiDiagnostic.Code.JSII_3001_EXPOSED_INTERNAL_TYPE.create(
+          typeAnnotationNode,
+          type.symbol,
+          isThisType,
+          typeUse,
+        ).addRelatedInformation(
+          JsiiDiagnostic.Code.JSII_9999_RELATED_INFO.create(
```

Comment:
`JSII_9999_RELATED_INFO `: this feels like cheating. Are "RelatedInformation"s *really* a type of Diagnostic?

Or do they just happen to share some features like formatting and a location and was this a convenient way to get that? Some remodeling might be in order.

Reply:
The `ts.DiagnosticRelatedInformation` interface is actually the base interface of `ts.Diagnostic` (the later only adds properties we don't care about using - `reportsUnnecessary`, `reportsDeprecated`, `source` - as well as the `relatedInformation` property that is an optional array of `ts.DiagnosticRelatedInformation`).

The consequence of this is that a `code` is needed on the related information in the exact same way that for top-level diagnostics... And those codes are actually visible to programmatic users.